test: complete impl. of msg_merkleblock and wait_for_merkleblock

theStack opened the pull request:
Implements the missing initialization/serialization methods for `msg_merkleblock`, based on the already present class `CMerkleBlock`. Also changes the method `wait_for_merkleblock()` to be more precise by waiting for a merkleblock with a specified blockhash instead of an arbitrary one.

In the BIP37 test `p2p_filter.py`, this new method is used to make the test of receiving merkleblock and tx if a filter is set to be more precise, by checking if they also arrive in the right order.

In the course of this PR, also the interface for the methods `wait_for_merkleblock()` and `wait_for_header()` are improved to take a hex string instead of an integer, which is more typesafe and less of a burden to the caller.

theStack committed:
test: complete impl. of msg_merkleblock and wait_for_merkleblock

Implements the missing initialization/serialization methods for
msg_merkleblock, based on the already present class CMerkleBlock.
Also changes the method wait_for_merkleblock() to be more precise by waiting
for a merkleblock with a specified blockhash instead of an arbitrary one.

In the BIP37 test p2p_filter.py, this new method is used to make the test of
receiving merkleblock and tx if a filter is set to be more precise, by checking
if they also arrive in the right order.

MarcoFalke commented on test/functional/p2p_filter.py:
```diff
@@ -69,18 +69,15 @@ def run_test(self):
-        self.nodes[0].generatetoaddress(1, self.nodes[0].getnewaddress())
-        filter_node.wait_for_merkleblock()
+        block_hash = self.nodes[0].generatetoaddress(1, self.nodes[0].getnewaddress())[0]
+        filter_node.wait_for_merkleblock(int(block_hash, 16))
```
style-nit: I slightly prefer the interface where the hash is passed in as a hex string, as opposed to a plain integer. The integer is less type safe and could be confused with other args such as timeout. Also, putting the burden on the caller to always call `int()` makes tests slightly harder to write.

theStack replied:
I fully agree, but then for consistency reasons all three methods `wait_for_{header,block,merkleblock}` should have the same interface and be changed to take a hex string accordingly? Maybe a good candidate for a follow-up PR (good first issue).

MarcoFalke replied:
At least wait_for_header is only used in one place:

```
$ git grep wait_for_header
test/functional/p2p_invalid_locator.py:                node.p2p.wait_for_header(int(node.getbestblockhash(), 16))
```

so this could be sneaked into this pull (as a separate commit) or left for later. Whatever you prefer.

maflcko reviewed (approved):
ACK, nice I didn't know `CMerkleBlock` was already implemented

theStack committed:
refactor: test: improve wait_for{header,merkleblock} interface

The interfaces for the methods wait_for_header() and wait_for_merkleblock() are
changed to take a hex string instead of an integer, improving type safety and
removing the burden from the caller to always do the transformation via
`int(...)`. As suggested by MarcoFalke in
https://github.com/bitcoin/bitcoin/pull/18593#discussion_r407062253

theStack commented:
Pushed another commit that improves the interface of `wait_for_merkleblock` and `wait_for_header` by taking a hex string instead of an integer, [as](https://github.com/bitcoin/bitcoin/pull/18593#discussion_r407062253) [suggested](https://github.com/bitcoin/bitcoin/pull/18593#discussion_r407064237).

MarcoFalke commented:
```
# ./test/lint/lint-python.sh 
test/functional/p2p_filter.py:17:1: F401 'test_framework.mininode.mininode_lock' imported but unused

theStack commented:
@MarcoFalke: Thanks for the hint, fixed. (Should also form the habit of running the linters locally before pushing...)

MarcoFalke commented:
No worries. The linter should run on travis in the first couple of seconds and then either fail or pass. I don't even have flake8 installed on my workstation.

MarcoFalke commented:
ACK 854382885f18aa9a95cdde3d11591b05c305ad3f
